### ISSUE TITLE
[Bug] The app crash with native exception when changing visibility of Grid

### ISSUE THREAD

**Issue by chamons** [mono/mono#20490]
_From @SeRgI1982 on Wed, 14 Oct 2020 12:43:33 GMT_

### Description
It happens very rarely but sometimes when I enter to the ContentPage the app crash. 

> ```
> 2020-10-14 14:20:15.185 MyApp[7056:102252] error: mono_coop_mutex_lock Cannot transition thread 0x112ea1dc0 from STATE_BLOCKING with DO_BLOCKING
> 
> =================================================================
> 	Native Crash Reporting
> =================================================================
> Got a abrt while executing native code. This usually indicates
> a fatal error in the mono runtime or one of the native libraries 
> used by your application.
> =================================================================
> 
> =================================================================
> 	Native stacktrace:
> =================================================================
> 	0x10a9a5b96 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_dump_native_crash_info
> 	0x10a999785 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_handle_native_crash
> 	0x10a9a519f - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : sigabrt_signal_handler
> 	0x7fff6d0b85fd - /usr/lib/system/libsystem_platform.dylib : _sigtramp
> 	0x112df8107 - Unknown
> 	0x7fff6cf8e808 - /usr/lib/system/libsystem_c.dylib : abort
> 	0x10a7fd7f0 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : _ZL12log_callbackPKcS0_S0_iPv
> 	0x10abc92b5 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : monoeg_g_logv_nofree
> 	0x10abc9372 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : monoeg_g_log
> 	0x10abba0dc - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_threads_transition_do_blocking
> 	0x10abb7863 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_threads_enter_gc_safe_region_unbalanced_with_info
> 	0x10aa0a3f8 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : suspend_current
> 	0x10aa08559 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : try_process_suspend
> 	0x10aa2545b - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_de_process_single_step
> 	0x10aa0761c - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : debugger_agent_single_step_from_context
> 	0x10dabbeff - Unknown
> 	0x11192ebbd - Unknown
> 	0x11192eb13 - Unknown
> 	0x119098a4b - Unknown
> 	0x1190987fb - Unknown
> 	0x116f0b8fb - Unknown
> 	0x116f0742b - Unknown
> 	0x116f053b3 - Unknown
> 	0x116f2be73 - Unknown
> 	0x1167f90b2 - Unknown
> 	0x116488d0c - Unknown
> 	0x1164ba1b3 - Unknown
> 	0x1164a86ef - Unknown
> 	0x1164b85cb - Unknown
> 	0x1164b782b - Unknown
> 	0x1164b73bb - Unknown
> 	0x11b685883 - Unknown
> 	0x13cd2efeb - Unknown
> 	0x13cd2e884 - Unknown
> 	0x1190c39ff - Unknown
> 	0x111bacebe - Unknown
> 	0x10a9b04be - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_jit_runtime_invoke
> 	0x10aae5558 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_runtime_invoke_checked
> 	0x10aae9e2e - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_runtime_invoke
> 	0x10ac60ce9 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : _ZL31native_to_managed_trampoline_29P11objc_objectP13objc_selectorPP11_MonoMethodj
> 	0x10ac6112c - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : -[__MonoMac_NSAsyncSynchronizationContextDispatcher xamarinApplySelector]
> 	0x7fff355240dd - /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation : __NSThreadPerformPerform
> 	0x7fff32e4fd52 - /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation : __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
> 	0x7fff32e4fcf1 - /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation : __CFRunLoopDoSource0
> 	0x7fff32e4fb0b - /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation : __CFRunLoopDoSources0
> 	0x7fff32e4e83a - /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation : __CFRunLoopRun
> 	0x7fff32e4de3e - /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation : CFRunLoopRunSpecific
> 	0x7fff31a7aabd - /System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/HIToolbox : RunCurrentEventLoopInMode
> 	0x7fff31a7a7d5 - /System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/HIToolbox : ReceiveNextEventCommon
> 	0x7fff31a7a579 - /System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/HIToolbox : _BlockUntilNextEventMatchingListInModeWithFilter
> 	0x7fff300c0039 - /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit : _DPSNextEvent
> 	0x7fff300be880 - /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit : -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:]
> 	0x7fff300b058e - /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit : -[NSApplication run]
> 	0x7fff30082396 - /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit : NSApplicationMain
> 	0x11523a0a7 - Unknown
> 	0x114f9dd23 - Unknown
> 	0x10db31923 - Unknown
> 	0x10db31e71 - Unknown
> 	0x10a9b04be - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_jit_runtime_invoke
> 	0x10aae5558 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_runtime_invoke_checked
> 	0x10aaec885 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_runtime_exec_main_checked
> 	0x10a90c92c - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_jit_exec
> 	0x10a90fa26 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : mono_main
> 	0x10a80a37c - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : xamarin_main
> 	0x10a80b1c4 - /Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MacOS/MyApp : main
> 	0x7fff6cebfcc9 - /usr/lib/system/libdyld.dylib : start
> 
> =================================================================
> 	Telemetry Dumper:
> =================================================================
> Pkilling 0x123145490747392x from 0x4612300224x
> ...
> Pkilling 0x123145484419072x from 0x4612300224x
> Entering thread summarizer pause from 0x4612300224x
> Could not exec mono-hang-watchdog, expected on path '/Users/MyUser/MyAppX/MyApp.Products.MyAppX.MacOS/bin/Debug/MyApp.app/Contents/MonoBundle/../bin/mono-hang-watchdog' (errno 2)
> Finished thread summarizer pause from 0x4612300224x.
> Failed to create breadcrumb file (null)/crash_hash_0x1237ee7fc4
> 
> Waiting for dumping threads to resume
> 
> =================================================================
> 	External Debugger Dump:
> =================================================================
> Mono.Debugger.Soft.VMDisconnectedException: Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
>  at Mono.Debugger.Soft.ThreadMirror.GetFrames () [0x00038] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugger.Soft/Mono.Debugger.Soft/ThreadMirror.cs:30 
>  at Mono.Debugging.Soft.SoftDebuggerSession.EvaluateExpression (Mono.Debugger.Soft.ThreadMirror thread, System.String expression, Mono.Debugging.Client.BreakEvent bp) [0x00000] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugging.Soft/SoftDebuggerSession.cs:2409 Mono.Debugger.Soft.VMDisconnectedException: Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
>  at Mono.Debugger.Soft.ThreadMirror.GetFrames () [0x00038] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugger.Soft/Mono.Debugger.Soft/ThreadMirror.cs:30 
>  at Mono.Debugging.Soft.SoftDebuggerSession.EvaluateExpression (Mono.Debugger.Soft.ThreadMirror thread, System.String expression, Mono.Debugging.Client.BreakEvent bp) [0x00000] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugging.Soft/SoftDebuggerSession.cs:2409 Mono.Debugger.Soft.VMDisconnectedException: Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
>  at Mono.Debugger.Soft.ThreadMirror.GetFrames () [0x00038] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugger.Soft/Mono.Debugger.Soft/ThreadMirror.cs:30 
>  at Mono.Debugging.Soft.SoftDebuggerSession.EvaluateExpression (Mono.Debugger.Soft.ThreadMirror thread, System.String expression, Mono.Debugging.Client.BreakEvent bp) [0x00000] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugging.Soft/SoftDebuggerSession.cs:2409 Mono.Debugger.Soft.VMDisconnectedException: Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
>  at Mono.Debugger.Soft.ThreadMirror.GetFrames () [0x00038] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugger.Soft/Mono.Debugger.Soft/ThreadMirror.cs:30 
>  at Mono.Debugging.Soft.SoftDebuggerSession.EvaluateExpression (Mono.Debugger.Soft.ThreadMirror thread, System.String expression, Mono.Debugging.Client.BreakEvent bp) [0x00000] in /Users/builder/azdo/_work/3/s/main/external/debugger-libs/Mono.Debugging.Soft/SoftDebuggerSession.cs:2409 [ALL V&E]= [IsSyncViewMaster]= [IsForcingForAll]= [IsActive]=
>
> =================================================================
> 	Basic Fault Address Reporting
> =================================================================
> Memory around native instruction pointer (0x7fff6d00733a):0x7fff6d00732a  ff ff c3 90 90 90 b8 48 01 00 02 49 89 ca 0f 05  .......H...I....
> 0x7fff6d00733a  73 08 48 89 c7 e9 e5 a2 ff ff c3 90 90 90 b8 53  s.H............S
> 0x7fff6d00734a  00 00 02 49 89 ca 0f 05 73 08 48 89 c7 e9 cd a2  ...I....s.H.....
> 0x7fff6d00735a  ff ff c3 90 90 90 b8 83 01 00 02 49 89 ca 0f 05  ...........I....
> 
> =================================================================
> 	Managed Stacktrace:
> =================================================================
>	  at <unknown> <0xffffffff>
>	  at CoreAnimation.CATransform3D:CATransform3DRotate <0x0017a>
>	  at CoreAnimation.CATransform3D:Rotate <0x0020a>
>	  at <>c__DisplayClass31_0:<OnUpdateNativeControl>g__update|0 <0x01dda>
>	  at Xamarin.Forms.Platform.MacOS.VisualElementTracker:OnUpdateNativeControl <0x015ca>
>	  at Xamarin.Forms.Platform.MacOS.VisualElementTracker:UpdateNativeControl <0x00352>
>	  at Xamarin.Forms.Platform.MacOS.VisualElementTracker:HandlePropertyChanged <0x00e72>
>	  at <Module>:invoke_void_object_PropertyChangedEventArgs <0x001c1>
>	  at Xamarin.Forms.BindableObject:OnPropertyChanged <0x0013b>
>	  at Xamarin.Forms.Element:OnPropertyChanged <0x000d2>
>	  at Xamarin.Forms.BindableObject:SetValueActual <0x0088e>
>	  at Xamarin.Forms.BindableObject:SetValueCore <0x00c5a>
>	  at Xamarin.Forms.BindableObject:SetValue <0x0041a>
>	  at Xamarin.Forms.BindableObject:SetValue <0x000aa>
>	  at Xamarin.Forms.VisualElement:set_IsVisible <0x000e2>
>	  at MyApp.Products.MyAppX.Views.MapPage:**OnToasNotificationRequested** <0x0072a>
>	  at <>c__DisplayClass2_0:<InvokeAction>b__0 <0x000d3>
>	  at Foundation.NSAsyncSynchronizationContextDispatcher:Apply <0x000ce>
>	  at System.Object:runtime_invoke_void__this__ <0x0010d>
>	  at <unknown> <0xffffffff>
>	  at AppKit.NSApplication:NSApplicationMain <0x00276>
>	  at AppKit.NSApplication:Main <0x00292>
>	  at MyApp.Products.MyAppX.macOS.MainClass:Main <0x001b2>
>	  at <Module>:runtime_invoke_void_object <0x00110>
> =================================================================
> ```
### OnToasNotificationRequested

```
notificationPanelContainer.IsVisible = true;
notificationPanelContainer.InputTransparent = true;
notificationPanelContainer.CascadeInputTransparent = true;
notificationPanel.TranslationX = -400;
notificationPanel.TranslationY = 0;
notificationPanel.SetMessage(e.Message);
notificationPanel.TranslateTo(0, 0, 750)
                 .ContinueWith(r =>
                 {
                     if (r.Exception == null && !e.IsIndeterminate)
                     {
                         _reloadTimer.Start();
                     }
                 });
```
### Where in the XAML it looks in that way

```
<!--  Notification panel  -->
<Grid x:Name="notificationPanelContainer"
      AbsoluteLayout.LayoutBounds="0,1,1,0.25"
      AbsoluteLayout.LayoutFlags="All"
      InputTransparent="True"
      IsVisible="False">
    <ctrl:NotifcationPanel x:Name="notificationPanel"
                           HorizontalOptions="Start"
                           InputTransparent="True"
                           VerticalOptions="End" />
</Grid> 
```

### Basic Information

- Version with issue: Xamarin Forms 4.8.0.1269
- Last known good version: it happened with previous versions too
- IDE: VS for Mac
- Platform Target Frameworks: 
  - macOS:  Catalina 10.15.7
- Nuget Packages:
- Affected Devices: MacBook Pro (Retina, 15-inch, Mid 2014) 2.2 GHz Quad-Core i7, 16 GB 1500 MHz DDR3, Intel Iris Pro 1536 MB

_Copied from original issue xamarin/xamarin-macios#9873_

**Comment by lambdageek** [comment]
Probably `debugger_agent_single_step_from_context` needs a transition to GC Unsafe mode - it looks like it got called while the thread was doing some native core animation code.

**Comment by lambdageek** [comment]
Fixed on master and on 6.12 (2020-02) should be in 6.12.0.103 or later
